fix: Mark task result as completed if pod has been deleted for a while. Fixes #13533

jswxstw opened the pull request:
<!-- markdownlint-disable MD041 -- this is rendered within existing HTML, so allow starting without an H1 -->

<!--

### Before you open your PR

- Run `make pre-commit -B` to fix codegen and lint problems (build will fail).
- [Signed-off your commits](https://github.com/apps/dco/) (otherwise the DCO check will fail).
- Used [a conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/).

### When you open your PR

- PR title format should also conform to [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).
- "Fixes #" is in both the PR title (for release notes) and this description (to automatically link and close the issue).
- Create the PR as draft.
- Once builds are green, mark your PR "Ready for review".

When changes are requested, please address them and then dismiss the review to get it reviewed again.

-->

<!-- Does this PR fix an issue -->

Fixes #13533

### Motivation

see #13533 for details.

### Modifications

Mark task result as completed if pod has been deleted for a while.

### Verification

Corner case, hard to reproduce.

<!--
### Beyond this PR

Thank you for submitting this! Have you ever thought of becoming a Reviewer or Approver on the project?

Argo Workflows is seeking more community involvement and ultimately more [Reviewers and Approvers](https://github.com/argoproj/argoproj/blob/main/community/membership.md) to help keep it viable.
See [Sustainability Effort](https://github.com/argoproj/argo-workflows/blob/main/community/sustainability_effort.md) for more information.

-->

jswxstw commented:
/retest

jswxstw commented:
/retest

agilgur5 commented on workflow/controller/taskresult.go:
```diff
@@ -54,19 +54,14 @@ func (wfc *WorkflowController) newWorkflowTaskResultInformer() cache.SharedIndex
-func podAbsentTimeout(node *wfv1.NodeStatus) bool {
-	return time.Since(node.StartedAt.Time) <= envutil.LookupEnvDurationOr("POD_ABSENT_TIMEOUT", 2*time.Minute)
+func recentlyDeleted(node *wfv1.NodeStatus) bool {
+	return time.Since(node.FinishedAt.Time) <= envutil.LookupEnvDurationOr("RECENTLY_DELETED_POD_DURATION", 10*time.Second)
```
If I'm not mistaken, the reason for using `startedAt` was because the `finishedAt` time may never have been recorded

jswxstw replied:
> this almost certainly breaks the part of https://github.com/argoproj/argo-workflows/pull/13454 which was to ensure 3.4->3.5 upgrade worked as hoped for in flight workflows.

Do you mean this problem https://github.com/argoproj/argo-workflows/issues/12103#issuecomment-2185487682?

isubasinghe replied:
That is correct. @jswxstw do you think you would be able to make time for a chat about this issue?

isubasinghe replied:
Actually this doesn't break the 3.4 -> 3.5 upgrade path, the else if check is present meaning we never falsely update the TaskCompletionStatus.

jswxstw replied:
> That is correct. @jswxstw do you think you would be able to make time for a chat about this issue?

Sure, I have time now.

> Actually this doesn't break the 3.4 -> 3.5 upgrade path, the else if check is present meaning we never falsely update the TaskCompletionStatus.

Yes, I kept this fix.

isubasinghe replied:
> Sure, I have time now.

Its alright, I have enough confidence now that this is a better approach than #13537

agilgur5 replied:
Related: https://github.com/argoproj/argo-workflows/pull/13798#discussion_r1818921785

isubasinghe reviewed (approved):
This code should fix #13533. 

I am unsure if this will fix all related issues however. 

Confirmed that in flight workflows from 3.4 -> 3.5 upgrade path will work as expected.

Unfortunately, this path is impacted by another bug and will result in infinite recursion. There is a fix for this as well.

jswxstw commented:
/retest

jswxstw commented:
> I am unsure if this will fix all related issues however.

@isubasinghe As far as I know, almost all the related issues are that wait container does not report `LabelKeyReportOutputsCompleted` as expected:
1. waiti container with old versions do not have the label `LabelKeyReportOutputsCompleted`, fixed by #13454
2. wait container exits abnormally, including the following two situations:
    - pod not deleted: fixed by #13491
    - pod deleted: fixed by current issue.

isubasinghe commented:
> > I am unsure if this will fix all related issues however.
> 
> @isubasinghe As far as I know, almost all the related issues are that wait container does not report `LabelKeyReportOutputsCompleted` as expected:
> 
>     1. wait container with old versions do not have the label `LabelKeyReportOutputsCompleted`, fixed by [fix: mark node failed if pod absent. Fixes #12993 #13454](https://github.com/argoproj/argo-workflows/pull/13454)
> 
>     2. wait container exits abnormally, including the following two situations:
>        
>        * pod not deleted: fixed by [fix: Mark taskResult completed if wait container terminated not gracefully. Fixes #13373 #13491](https://github.com/argoproj/argo-workflows/pull/13491)
>        * pod deleted: fixed by current PR.

Makes sense, I've gone through the code in quite a lot of depth for all of today effectively and I came to a similar conclusion, your agreement clears my doubts, thanks.

ericblackburn commented:
I've upgraded to 3.5.11, [using chart 0.42.3](https://artifacthub.io/packages/helm/argo/argo-workflows), to use this bug fix but still see that workflows are stuck in a running state
![image](https://github.com/user-attachments/assets/70a34372-eaed-4ffc-bc7d-848253a608d3)
![image](https://github.com/user-attachments/assets/4fd821db-148e-4aa0-a2ed-2aaed89e1b3a)

jswxstw commented:
> time="2024-09-24T16:26:00.429Z" level=warning msg="workflow uses legacy/insecure pod patch, see https://argo-workflows.readthedocs.io/en/release-3.5/workflow-rbac/" namespace=testnamespace workflow=testworkflow-1727192700-n4v9v

@ericblackburn I think you should fix the RBAC problem.

ericblackburn commented:
[Yep, that was it.](https://github.com/argoproj/argo-workflows/pull/13537#issuecomment-2378303180) Good call out.
![image](https://github.com/user-attachments/assets/2ae5d05f-f5ea-47d6-a646-b70fdf9c1168)

isubasinghe commented:
@jswxstw unfortunately I can confirm that the node.Message can be something other than  "pod deleted" when a pod is killed, this means that your fix won't work. Do you have any  ideas about how we can tackle this without checking the message?

jswxstw commented:
> @jswxstw unfortunately I can confirm that the node.Message can be something other than "pod deleted" when a pod is killed, this means that your fix won't work. Do you have any ideas about how we can tackle this without checking the message?

Do you mean a pod will be deleted with a message not "pod deleted"?
If a pod is just killed but not deleted yet, #13491 will check the actual status of `wait` container.

isubasinghe commented:
@jswxstw sorry yeah I meant deleted

isubasinghe commented:
I think we should just see if the pod for that node disappeared in the `IsPodDeleted` function, opinions @jswxstw ?

jswxstw commented:
> I think we should just see if the pod for that node disappeared in the `IsPodDeleted` function, opinions @jswxstw ?

Yes, we can use `woc.podExists` instead to check if its pod has disappeared.
However, If a pod is deleted with a message not "pod deleted", it means controller has observed the pod status after `podReconciliation` and it should already know the actual status of `wait` container. Are there other scenarios I haven't considered?

Joibel commented:
I would like it to wait some time between observing a gone pod without a `completed` workflow task result and us kicking in the marking of the task result as completed. It could be that the completed WFTresult just hasn't been observed yet.

jswxstw commented:
> This might be due to a race condition, I am currently unsure of how this happens.

https://github.com/argoproj/argo-workflows/blob/cf6223d8f21e1cfb49492c6b5aeca40752ee2787/workflow/controller/operator.go#L1459-L1461

Under normal circumstances, the node can only be set to `Completed` when `wait` container has truly exited.
Therefore, in theory, we should always be able to obtain the actual state of `wait` container here, unless the pod is deleted before the controller has a chance to observe it. In such cases, it will be marked as 'pod deleted'.

isubasinghe commented:
Okay I found out what happens. If the node has been marked failed by some other failure and we are still waiting on task results, the message will never be updated to reflect `pod deleted` since podReconciliation will skip this node. 

I suggest that markNodePhase should be responsible for patching marking the task result completed in this case.
I see no reason why we would want to wait around for a task result when the node has already been marked Failed/Errored?

isubasinghe commented:
Opinions @jswxstw ?

jswxstw commented:
> Okay I found out what happens. If the node has been marked failed by some other failure and we are still waiting on task results, the message will never be updated to reflect `pod deleted` since podReconciliation will skip this node.

@isubasinghe Do you mean the node is marked failed in `podReconciliation`? It will not bring any problems if so because controller will check the actual state of `wait` container.

I found that it will casue some problems if controller marks the node as failed directly regardless of the pod's status, see https://github.com/argoproj/argo-workflows/issues/13726#issuecomment-2399634859. Are you encountering similar issues?

isubasinghe commented:
@jswxstw I mean  if it is marked elsewhere. 

I think `markNodePhase` should include an if condition to check if the phase is error/failure and then markTaskResultsComplete(node.ID)

jswxstw commented:
> @jswxstw I mean if it is marked failed elsewhere, not in podReconciliation.

@jswxstw Could you specify exactly where? In my opinion controller should not mark the node as failed directly regardless of the pod's status.

isubasinghe commented:
@jswxstw I wasn't able to find where exactly, I think what we should do is if we mark nodes failed, we should mark the task result complete.

Also if we get a false label in `taskResultReconciliation` we should first make sure the node hasn't failed or errored first, before labelling task result incomplete.

jswxstw commented:
@isubasinghe I think there are risks in doing this, because at this point we cannot confirm whether the informer has watched task result updates, this process may be delayed.

isubasinghe commented:
@jswxstw but we don't care about a task result after we fail a node, correct? So it doesn't matter if we receive a task result or not.

jswxstw commented:
@isubasinghe We may need the exact exit code of the node from the task result.

isubasinghe commented:
@jswxstw I think it should be okay after looking at the code.

We seem to only add to scope if the node succeeded: https://github.com/argoproj/argo-workflows/blob/bb5130e841ab02cfe5b65f19c214fe309ca88bce/workflow/controller/operator.go#L1144 

So you are right that the second conditional will now return true, but he first condition still returns false. Therefore this behaviour should remain the same.

isubasinghe commented:
Steps/DAGs doesn't seem to care about task results. I wonder if this is a bug on that part of  the codebase ?

jswxstw commented:
> So you are right that the second conditional will now return true, but he first condition still returns false. Therefore this behaviour should remain the same.

There is an inconsistency in the judgment logic in these two places:
https://github.com/argoproj/argo-workflows/blob/bb5130e841ab02cfe5b65f19c214fe309ca88bce/workflow/controller/operator.go#L2335

https://github.com/argoproj/argo-workflows/blob/bb5130e841ab02cfe5b65f19c214fe309ca88bce/workflow/controller/operator.go#L1144

> Steps/DAGs doesn't seem to care about task results. I wonder if this is a bug on that part of the codebase ?

Only nodes of pod type have task results, other nodes depend on their child nodes.

isubasinghe commented:
In my PR I have proposed an alternative fix: https://github.com/argoproj/argo-workflows/pull/13798, 
would you still prefer this?

jswxstw commented:
> Sorry I meant, pods created by the templates {Step, DAG}. They do not wait until a task result is complete before creating a child. I think this is a bug? Because a task could have a `continueOn: Failed` and the child could rely upon the parents exit code.

Agree, we may also check if task result is Incomplete when node is failed.

> would you still prefer this?

This PR may exacerbate the above bug, but it is likely to have no impact, as nodes are rarely marked as failed directly outside of `podReconciliation`(this only occurs in the scenario of workflow shutdown).

isubasinghe commented:
> but it is likely to have no impact 

yeah this is what I think as well. 


I do think we need to rethink the task result implementation entirely.

isubasinghe commented:
@jswxstw I made the change here https://github.com/argoproj/argo-workflows/pull/13798 to check if the pod exists instead.

It turns out we do care about failed nodes, this was why the test was failing. 

I'm not sure if you'd agree with me, but I now think that this task result should maybe even be communicated via grpc or something like that to the controller directly. 
I

agilgur5 commented:
> I'm not sure if you'd agree with me, but I now think that this task result should maybe even be communicated via grpc or something like that to the controller directly.

security-wise, the Executor should never be able to directly communicate with the Controller. The CRs partly exist for that purpose, as well as for (temporary) durable storage in case of an outage

isubasinghe commented:
> Security-wise, the Executor should never need to nor be able to directly communicate with the Controller.

I don't really risk here though personally, could you elaborate how this creates a security risk. 
I mean I don't really expect to make this change, but we want consistency. 


We only have two choices here that I would be confident fixes this class of bugs:

a) communicate directly to the controller.
b)  bypass the cache and hit the API directly

agilgur5 commented:
> I don't really see the risk here though personally, could you elaborate how this creates a security risk.

The Controller would now be accessible via network from every namespace that Workflows runs in. Other than breaking network policies all over, this creates a whole new attack surface, as well as gives the Controller an API as an additional attack surface (and dependencies).
Anything would be able to hit that API as well, since the Executor is like any other Workflow Pod container; comparatively RBAC is used to control access to `WorkflowTaskResults` etc

That's a massive can of worms and fundamental shift in the architecture that I don't think should be changed.

jswxstw commented:
> I'm not sure if you'd agree with me, but I now think that this task result should maybe even be communicated via grpc or something like that to the controller directly.

I agree with @agilgur5. Argo, as a workflow engine for Kubernetes, is implemented based on asynchronous coordination. Transforming it into a synchronous mode is not in line with the Kubernetes style.

isubasinghe commented:
> The Controller would now be accessible via network from every namespace that Workflows runs in. 

Yeah that is fair enough. 

> That's a massive can of worms and fundamental shift in the architecture that I don't think should be changed.

Yeah I agree on this, I just don't see how to fix these issue with any sort of confidence.

jswxstw commented:
I feel that the reasons for these issues are twofold: 
- one is that we have split the original atomic operation to two: `podReconciliation` and `taskResultReconciliation`
- the other is that we did not fully consider the scenarios involving the coordination order and the loss of resources to be coordinated.

After so many fixes, we have actually covered the vast majority of scenarios. If there are still any edge cases that haven't been considered, the probability of them occurring will be very low, and the risk is manageable.

agilgur5 commented:
> state tracking can be done with a database

There's very little difference between a DB and using k8s resources/etcd. That also doesn't solve Pod tracking, which already uses k8s.
And similarly to my comments [above](https://github.com/argoproj/argo-workflows/pull/13537#issuecomment-2436694363) about Executor -> Controller connections, Executor -> DB connections also open up a can of worms around security

agilgur5 commented:
> I feel that the reasons for these issues are twofold:

Yea that sounds about right to me, with non-graceful termination as part of 2. 1. can't be done atomically given the different Informers involved, so it needs a different form of async coordination with some assumptions baked in
